### ISSUE TITLE
Possible issue with directedness of ECR gate causing invalid compiled circuits.

### ISSUE THREAD

**Issue by CalMacCQ** [CQCL/pytket-qiskit#312]
Using pytket-qiskit == 0.51.0 and pytket==1.26.0

```python
from pytket import Circuit
from pytket.extensions.qiskit import IBMQBackend

ghz3 = Circuit(3)
ghz3.H(0).CX(0, 1).CX(0, 2)
ghz3.measure_all()

ibm_backend = IBMQBackend("ibm_kyoto")

compiled_ghz3 = ibm_backend.get_compiled_circuit(ghz3)
```

If I do 

```python
ibm_backend.valid_circuit(compiled_ghz3)
```
I get `True` implying that the circuit is valid for the Kyoto architecture.

However if I execute the circuit I get the following error

```python
handle = ibm_backend.process_circuit(compiled_ghz3, n_shots=1000)
```

IBMInputValueError: 'The instruction ecr on qubits (12, 17) is not supported by the target system. Circuits that do not match the target hardware definition are no longer supported after March 4, 2024. See the transpilation documentation (https://docs.quantum.ibm.com/transpile) for instructions to transform circuits and the primitive examples (https://docs.quantum.ibm.com/run/primitives-examples) to see this coupled with operator transformations.'

```python
ibm_backend.backend_info.all_edge_gate_errors
```
This lists (17, 12) as a valid interaction but not (12, 17)

```
(node[17], node[12]): {<OpType.ECR: 66>: 0.003961009499739737}
```

This is concerning. We should ensure that the directedness of the ECR gate is accounted for to ensure compiled circuits are executable on the device.

https://github.com/Qiskit/qiskit-ibm-runtime/issues/1462

@yoshi-qc

**Comment by yoshi-qc** [comment]
`ECR(1,0)` can be described using `ECR(0,1)` as follows:
```
ecr10 = Circuit(2)
ecr10.Rz(0.5,0).SX(0).Rz(0.5,0).X(0).Rz(-0.5,0)
ecr10.Rz(0.5,1).SX(1).Rz(1.5,1).SX(1)
ecr10.ECR(0,1)
ecr10.Rz(0.5,1).SX(1)
ecr10.SX(0).Rz(0.5,0)
ecr10.add_phase(1.25)
```
And, for Eagle devices (here `ibm_backend = IBMQBackend("ibm_osaka")`), we create the direction list of ECR gates as follows:
`[i[1] for i in ibm_backend._backend_info.misc['characterisation']['GateTimes'] if i[0] == 'ecr']`.
The idea is that if we find that your compiled circuit by `get_compiled_circuit` contains an ECR gate whose direction isn't in this list, we can replace the ECR in ECR with another direction as above and get the circuit that Eagle devices can accept.

**Comment by yoshi-qc** [comment]
This compilation will not change the number of two-qubit gates, but it will increase the number of single-qubit gates, so there will be a chance to reduce the number of single-qubit gates.

**Comment by CalMacCQ** [comment]
Hi @yoshi-qc and @erinaldiq .... I have tried this again and the circuit now seems to execute as expected on the Kyoto architecture.

Could you try again and let me know if it works now? Maybe it was a temporary outage on the IBM side.

**Comment by yoshi-qc** [comment]
Hi @CalMacCQ Thank you for checking this issue.
For the GHZ state for 5 qubits, the compiled circuit with Eagle devices still does not work.
```
ghz5 = Circuit(5)
ghz5.H(0).CX(0,1).CX(0,2).CX(0,3).CX(0,4)
ghz5.measure_all()
ibm_ghz5 = ibm_backend.get_compiled_circuit(ghz5)
```

**Comment by CalMacCQ** [comment]
If indeed it is an issue with directedness of the ECR gate I am very suprised this hasn't shown up until now... Support for the ECR gate devices was added last summer in pytket-qiskit v0.40.0 and this has never shown up before. 

Worth resolving the cause of this before the 0.52.0 release.

**Comment by cqc-alec** [comment]
I'm unable to reproduce this.

```python
handle = ibm_backend.process_circuit(compiled_ghz3, n_shots=1000)
```

returns without error.

Maybe something non-deterministic is going on?

**Comment by yoshi-qc** [comment]
@cqc-alec Could you try the ghz of 5 qubits? I think ECR gates in your `compiled_ghz3` are implemented directions.

**Comment by cqc-alec** [comment]
> @cqc-alec Could you try the ghz of 5 qubits? I think ECR gates in your `compiled_ghz3` are implemented directions.

OK yes, with a 5-qubit circuit

```python
ghz.H(0).CX(0, 1).CX(0, 2).CX(0, 3).CX(0, 4)
```

I do get an error:

```
IBMInputValueError: 'The instruction ecr on qubits (35, 47) is not supported by the target system. Circuits that do not match the target hardware definition are no longer supported after March 4, 2024. See the transpilation documentation (https://docs.quantum.ibm.com/transpile) for instructions to transform circuits and the primitive examples (https://docs.quantum.ibm.com/run/primitives-examples) to see this coupled with operator transformations.'
```

**Comment by steffenbackes** [comment]
I can trigger the same 'ecr instruction is not supported on Qubits ...'  on ibm_kawasaki for the circuit
![image](https://github.com/CQCL/pytket-qiskit/assets/8615402/a2831f19-45d8-4339-b265-73efd6fb492a)
which gets compiled into this
![image](https://github.com/CQCL/pytket-qiskit/assets/8615402/41e61411-71b2-4d28-b8c7-d8df9d9e3c84)
and then running 
`expval = get_operator_expectation_value(circ, Op, backend, 5000,  partition_strat = PauliPartitionStrat.CommutingSets)`
for this operator
`-1.0×IIII +1.0×ZZII -0.5×XZXI -0.5×YZYI -0.5×IXZX -0.5×IYZY`

But without the partitioning 
`expval = get_operator_expectation_value(circ, Op, backend, 5000 )`
the circuit is accepted and pushed to the queue.

@yoshi-qc kindly provided me with a script to replace the ECR gate in the compiled circuit, it finds the ECR gate, but does not enter the condition for replacing them. 
So the error is still triggered, as long as 'PauliPartitionStrat.CommutingSets' is active.

**Comment by cqc-melf** [comment]
Solved and released, please update to `pytket-qiksit==0.52.0` 

https://github.com/CQCL/pytket-qiskit/releases/tag/v0.52.0
